src/filesystem.cpp

```diff
@@ -535,3 +539,29 @@ std::string ensure_valid_file_name( const std::string &file_name )
 
     return new_file_name;
 }
+
+bool is_lexically_valid( const fs::path& path ) {
+#if defined(_WIN32)
+    // Windows has strict rules for file naming
+    fs::path valid = path.root_path();
+    fs::path rel = path.relative_path();
+    // "Do not end a file or directory name with a space or a period."
+    // https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions
+    for( auto &it : rel ) {
+        std::string item = it.generic_u8string();
+        for( auto ch = item.rbegin(); ch != item.rend(); ++ch) {
+            if( *ch == ' ' || *ch == '.' ) {
+                return false;
+            } else {
+                break;
+            }
+        }
+        for( auto &it : item ) {
+            if( invalid_chars.find( it ) != std::string::npos ) {
+                return false;
+            }
+        }
+    }
+#endif
+    return true;
+}
```

Comment:
I think this is a nice chance to also check for the reserved Windows filenames, quoting [Microsoft documentation](https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file):

> Do not use the following reserved names for the name of a file:

> CON, PRN, AUX, NUL, COM0, COM1, COM2, COM3, COM4, COM5, COM6, COM7, COM8, COM9, COM¹, COM², COM³, LPT0, LPT1, LPT2, LPT3, LPT4, LPT5, LPT6, LPT7, LPT8, LPT9, LPT¹, LPT², and LPT³. Also avoid these names followed immediately by an extension; for example, NUL.txt and NUL.tar.gz are both equivalent to NUL.

So file or directory names equal to any of the above names or starting with any of the above names and a period should be invalid.